tree/tree/src/TTree.cxx

```diff
@@ -4457,96 +4460,121 @@ Int_t TTree::Fill()
    }
 #endif
 
-   if (fBranchRef) {
+   if (fBranchRef)
       fBranchRef->Fill();
-   }
+
    ++fEntries;
-   if (fEntries > fMaxEntries) {
+
+   if (fEntries > fMaxEntries)
       KeepCircular();
-   }
-   if (gDebug > 0) Info("TTree::Fill", " - A:  %d %lld %lld %lld %lld %lld %lld \n",
-       nbytes, fEntries, fAutoFlush,fAutoSave,GetZipBytes(),fFlushedBytes,fSavedBytes);
+
+   if (gDebug > 0)
+      Info("TTree::Fill", " - A: %d %lld %lld %lld %lld %lld %lld \n", nbytes, fEntries, fAutoFlush, fAutoSave,
+           GetZipBytes(), fFlushedBytes, fSavedBytes);
+
+   bool autoSave = false;
+   bool autoFlush = false;
 
    if (fAutoFlush != 0 || fAutoSave != 0) {
-      // Is it time to flush or autosave baskets?
       if (fFlushedBytes == 0) {
+         // If fFlushedBytes == 0, it means we never flushed or saved, so
+         // we need to check if it's time to do it and recompute the values
+         // of fAutoFlush and fAutoSave in terms of the number of entries.
          // Decision can be based initially either on the number of bytes
          // or the number of entries written.
+
          Long64_t zipBytes = GetZipBytes();
-         if ((fAutoFlush<0 && zipBytes > -fAutoFlush)  ||
-             (fAutoSave <0 && zipBytes > -fAutoSave )  ||
-             (fAutoFlush>0 && fEntries%TMath::Max((Long64_t)1,fAutoFlush) == 0) ||
-             (fAutoSave >0 && fEntries%TMath::Max((Long64_t)1,fAutoSave)  == 0) ) {
 
-            //First call FlushBasket to make sure that fTotBytes is up to date.
+         autoSave = fAutoSave < 0 ? (zipBytes > -fAutoSave) : fEntries == fAutoSave;
+         autoFlush = fAutoFlush < 0 ? (zipBytes > -fAutoFlush) : fEntries == fAutoFlush;
```

Comment:
One case where the new code is different than the old code:
```
t->SetAutoFlush(- 300 000 000 );
t->SetAutoSave(- 300 000 000 );
for(auto i = 0; i < 15; ++i) t->Fill(); // not enough for 300MB
t->SetAutoFlush(10);
// more filling.
```
In this case the old code would start autoflushing at entry 20 (fFlushedBytes==0 and fEntries%TMath::Max((Long64_t)1,fAutoFlush) == 0)
While the new code would only start autoflushing once one of the buffer 'overflows' (i.e. fFlushedBytes becomes non zero by other means).

Reply:
You are totally right, I'll drop the commit with this particular change.